### PR TITLE
issue #4713 New Resource: aws_neptune_cluster

### LINKED ISSUE #4713: Support Neptune
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### New or Affected Resource(s)

* aws_neptune_cluster
* aws_neptune_cluster_instance
* aws_neptune_subnet_group
* aws_neptune_parameter_group
* aws_neptune_cluster_parameter_group
* aws_neptune_event_subscription

### Potential Terraform Configuration

```hcl
resource "aws_neptune_cluster" "example" {
  cluster_identifier = "example"
  engine = "neptune"
  engine_version = "1.0.1"
  cluster_parameter_group_name = "${aws_neptune_cluster_parameter_group.example.name}"
  subnet_group_name = "${aws_neptune_subnet_group.example.name}"
  database_name = "database"
  master_username = "user"
  master_password = "password"
  ...

}

resource "aws_neptune_cluster_instance" "example" {
  identifier = "example"
  cluster_identifier = "${aws_neptune_cluster.example.id}"
  engine = "neptune"
  instance_class = "db.m4.large"
  ...

}

resource "aws_neptune_subnet_group" "example" {
  name = "example"
  description = "example description"
  subnet_ids = ["${aws_subnet.frontend.id}", "${aws_subnet.backend.id}"]
}

resource "aws_neptune_parameter_group" "example" {
  name = "example"
  family = "example"
  description = "example description"
}

resource "aws_neptune_cluster_parameter_group" "example" {
  name = "example"
  family = "example"
  description = "example description"
}

resource "aws_neptune_event_subscription" "example" {
  name = "example"
  sns_topic = "${aws_sns_topic.example.arn}"
  source_type = "db-instance"
  source_ids = ["${aws_neptune_cluster_instance.example.id}"]
  event_categories = [
    "creation",
    "failover",
    ...
  ]
}
```

### Order

1. `aws_neptune_parameter_group`,`aws_neptune_cluster_parameter_group`,`aws_neptune_subnet_group`
2. `aws_neptune_cluster`
3. `aws_neptune_cluster_instance`
4. `aws_neptune_event_subscription`

### PR DESCRIPTION
**Pull request by saravanan30erd**
~~Fixes #4713~~
Reference: #4713 

```
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSNeptuneCluster'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSNeptuneCluster -timeout 120m
=== RUN   TestAccAWSNeptuneClusterParameterGroup_basic
--- PASS: TestAccAWSNeptuneClusterParameterGroup_basic (55.86s)
=== RUN   TestAccAWSNeptuneClusterParameterGroup_namePrefix
--- PASS: TestAccAWSNeptuneClusterParameterGroup_namePrefix (54.48s)
=== RUN   TestAccAWSNeptuneClusterParameterGroup_generatedName
--- PASS: TestAccAWSNeptuneClusterParameterGroup_generatedName (55.00s)
=== RUN   TestAccAWSNeptuneClusterParameterGroup_Description
--- PASS: TestAccAWSNeptuneClusterParameterGroup_Description (55.70s)
=== RUN   TestAccAWSNeptuneClusterParameterGroup_Parameter
--- PASS: TestAccAWSNeptuneClusterParameterGroup_Parameter (94.20s)
=== RUN   TestAccAWSNeptuneClusterParameterGroup_Tags
--- PASS: TestAccAWSNeptuneClusterParameterGroup_Tags (134.52s)
=== RUN   TestAccAWSNeptuneCluster_basic
--- PASS: TestAccAWSNeptuneCluster_basic (130.39s)
=== RUN   TestAccAWSNeptuneCluster_namePrefix
--- PASS: TestAccAWSNeptuneCluster_namePrefix (152.58s)
=== RUN   TestAccAWSNeptuneCluster_takeFinalSnapshot
--- PASS: TestAccAWSNeptuneCluster_takeFinalSnapshot (212.69s)
=== RUN   TestAccAWSNeptuneCluster_updateTags
--- PASS: TestAccAWSNeptuneCluster_updateTags (177.25s)
=== RUN   TestAccAWSNeptuneCluster_updateIamRoles
--- PASS: TestAccAWSNeptuneCluster_updateIamRoles (219.52s)
=== RUN   TestAccAWSNeptuneCluster_kmsKey
--- PASS: TestAccAWSNeptuneCluster_kmsKey (196.87s)
=== RUN   TestAccAWSNeptuneCluster_encrypted
--- PASS: TestAccAWSNeptuneCluster_encrypted (142.91s)
=== RUN   TestAccAWSNeptuneCluster_backupsUpdate
--- PASS: TestAccAWSNeptuneCluster_backupsUpdate (199.69s)
=== RUN   TestAccAWSNeptuneCluster_iamAuth
--- PASS: TestAccAWSNeptuneCluster_iamAuth (130.87s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	2012.587s
...
```

### CODE REVIEW

**Comment by austinkelleher** [comment]
Awesome! We are using Neptune and would really love to see this get merged! Thank you for all of your work on this @saravanan30erd!

File: website/docs/r/neptune_cluster.html.markdown [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+```
+
+~> **Note:** AWS Neptune does not support user name/password–based access control.
+See the AWS [Docs](https://docs.aws.amazon.com/neptune/latest/userguide/limits.html) for more information.
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `cluster_identifier` - (Optional, Forces new resources) The cluster identifier. If omitted, Terraform will assign a random, unique identifier.
```
Comment (bflad) [inline website/docs/r/neptune_cluster.html.markdown]: The arguments should be alphabetically sorted, they are fairly hard to follow with the current ordering.

File: website/docs/r/neptune_cluster.html.markdown [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The Neptune Cluster Identifier
+* `cluster_identifier` - The Neptune Cluster Identifier
+* `cluster_resource_id` - The Neptune Cluster Resource ID
+* `cluster_members` – List of Neptune Instances that are a part of this cluster
+* `availability_zones` - The availability zone of the instance
```
Comment (bflad) [inline website/docs/r/neptune_cluster.html.markdown]: The attributes byline states `In addition to all arguments above` -- we should remove all arguments from this attributes list and alphabetically sort the remainder

**Comment by saravanan30erd** [comment]
@bflad updated the recommended changes 👍 

$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSNeptuneCluster_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSNeptuneCluster_ -timeout 120m
=== RUN   TestAccAWSNeptuneCluster_basic
--- PASS: TestAccAWSNeptuneCluster_basic (128.93s)
=== RUN   TestAccAWSNeptuneCluster_namePrefix
--- PASS: TestAccAWSNeptuneCluster_namePrefix (139.78s)
=== RUN   TestAccAWSNeptuneCluster_takeFinalSnapshot
--- PASS: TestAccAWSNeptuneCluster_takeFinalSnapshot (152.37s)
=== RUN   TestAccAWSNeptuneCluster_updateTags
--- PASS: TestAccAWSNeptuneCluster_updateTags (177.87s)
=== RUN   TestAccAWSNeptuneCluster_updateIamRoles
--- PASS: TestAccAWSNeptuneCluster_updateIamRoles (212.79s)
=== RUN   TestAccAWSNeptuneCluster_kmsKey
--- PASS: TestAccAWSNeptuneCluster_kmsKey (192.41s)
=== RUN   TestAccAWSNeptuneCluster_encrypted
--- PASS: TestAccAWSNeptuneCluster_encrypted (138.92s)
=== RUN   TestAccAWSNeptuneCluster_backupsUpdate
--- PASS: TestAccAWSNeptuneCluster_backupsUpdate (167.89s)
=== RUN   TestAccAWSNeptuneCluster_iamAuth
--- PASS: TestAccAWSNeptuneCluster_iamAuth (149.48s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1460.477s

File: aws/resource_aws_neptune_cluster.go [hunk 3]
```diff
@@ -0,0 +1,827 @@
...
+	}
+}
+
+func resourceAwsNeptuneClusterCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).neptuneconn
+	tags := tagsFromMapNeptune(d.Get("tags").(map[string]interface{}))
+
+	// Check if any of the parameters that require a cluster modification after creation are set
+	var clusterUpdate bool
+	clusterUpdate = false
```
Comment (bflad) [inline aws/resource_aws_neptune_cluster.go]: Minor nitpick: Boolean zero values are `false` so this line is extraneous. Reference: https://tour.golang.org/basics/12

Otherwise this could be declared as: `clusterUpdate := false`

**Comment by bflad** [comment]
This has been released in [version 1.28.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1280-july-18-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
